src/main/ruby/truffleruby/core/truffle/feature_loader.rb

```diff
@@ -140,6 +143,16 @@ def self.find_feature_or_file(feature, use_feature_provided = true)
       end
     end
 
+    def self.load_unless_realpath_loaded(feature, path)
+      # TODO: does this need to be synchronized?
```

Comment:
Yes of course, you could use a `TruffleRuby::ConcurrentMap`.
But I think this code would need to move to Java anyway, because we cannot have an extra backtrace entry for `require`, everything must happen in `Primitive.load_feature` (or another Primitive reusing it).